fenicsadapter/fenicsadapter.py

```diff
@@ -1,256 +1,35 @@
-                (self.is_vector_valued() and self._vals.ndim == self._dimension))
-
-        return_value = self._vals.ndim * [None]
-
-        if self._dimension == 1:
-            for i in range(self._vals.ndim):
-                return_value[i] = self._f[i](x[0])
-        if self._dimension == 2:
-            for i in range(self._vals.ndim):
-                return_value[i] = self._f[i](x[0], x[1])
-        if self._dimension == 3:
-            for i in range(self._vals.ndim):
-                return_value[i] = self._f[i](x[0], x[1], x[2])
-        return return_value
-
-
-class ExactInterpolationExpression(CustomExpression):
-    """Uses cubic spline interpolation for implementation of CustomExpression.interpolate. Only allows interpolation on
-    coupling that are parallel to the y axis, and if the coordinates in self._coords_y are ordered such that the nodes
-    on the coupling mesh are traversed w.r.t their connectivity.
-    However, this method allows to exactly recover the solution at the coupling interface, if it is a polynomial of
-    order 3 or lower.
-    See also https://github.com/precice/fenics-adapter/milestone/1
-    """
-    def create_interpolant(self):
-        interpolant = []
-        if self._dimension == 2:
-            if self.is_scalar_valued():  # check if scalar or vector-valued
-                interpolant.append(interp1d(self._coords_y, self._vals, bounds_error=False, fill_value="extrapolate", kind="cubic"))
-            elif self.is_vector_valued():
-                interpolant.append(interp1d(self._coords_y, self._vals[:, 0].flatten(), bounds_error=False, fill_value="extrapolate", kind="cubic"))
-                interpolant.append(interp1d(self._coords_y, self._vals[:, 1].flatten(), bounds_error=False, fill_value="extrapolate", kind="cubic"))
-            else:
-                raise Exception("Problem dimension and data dimension not matching.")
-        else:
-            raise Exception("Dimension of the function is invalid/not supported.")
 
-        return interpolant
-
-    def interpolate(self, x):
-        assert ((self.is_scalar_valued() and self._vals.ndim == 1) or
-                (self.is_vector_valued() and self._vals.ndim == self._dimension))
-
-        return_value = self._vals.ndim * [None]
-
-        if self._dimension == 2:
-            for i in range(self._vals.ndim):
-                return_value[i] = self._f[i](x[1])
-        else:
-            raise Exception("invalid dimensionality!")
-        return return_value
-
-
-class Adapter:
-    """Initializes the Adapter. Initalizer creates object of class Config (from
-    config.py module).
-
-    :ivar _config: object of class Config, which stores data from the JSON config file
-    """
-    def __init__(self, adapter_config_filename='precice-adapter-config.json',
-                 interpolation_strategy=GeneralInterpolationExpression):
+    def __init__(self, adapter_config_filename='precice-adapter-config.json'):
```

Comment:
I am not so happy with the way how the constructor works: The Object has many members that should be declared in `__init__`. However many of these members have to be initialized as `None`, since we cannot give them a reasonable value yet.

My proposal: Merge the functions `initialize(...)` and `set_interpolation_type(...)` into `__init__`. This will allow us to directly initialize the members of the object. If there is no use-case where thing have to be done after `__init__` but before `initialize(...)` is called. We can simply merge the functions. Similar to https://github.com/precice/precice/pull/614.

However, @uekerman might want to comment on this, since it will eliminate the `initialize(...)` function from the adapter, even though it is an important part of the preCICE API that we might want to keep for consistency reasons.

Reply:
Maybe we could even add (a part of) `initialize_data(...)`? Is it possible to use the adapter *without* calling `initialize_data`? preCICE allows this. Therefore, our adapter should also allow this.

I am not sure about the current restrictions of the adapter. Can we simply not call `initialize_data` and automatically initialize with zeros? I assume some additional information (function space...) might be missing.

Reply:
* All the members which are initialized as `None` are used more than once in the Adapter (mostly they are used in every time step). The general strategy followed is that if a member is used only once in an adapter function then it is defined locally within the scope of that function. **The essential trade-off here is to have less adapter member functions and increase preCICE API calls.**
* Currently the user needs to call `initialize_data()` without which the adapter cannot work. I will merge `initialize_data()` into `initialize()` in a way that it is not mandatory to call `initialize_data()` in the workflow

Reply:
With reference to [this study above](https://github.com/precice/fenics-adapter/pull/59#discussion_r422908918) API related member variables of `__init__ `will be replaced by API calls directly.